### PR TITLE
[JEP-216] Support loading localizer and webapp resources from any plugin

### PR DESCRIPTION
**Pull request by daniel-beck**
* Implement new Stapler SPI allowing override of resource lookup. Use that in turn to allow plugins to implement the same SPI as Jenkins `@Extension`.

Downstream https://github.com/stapler/stapler/pull/156, upstream from https://github.com/daniel-beck/localization-support-plugin and https://github.com/jenkinsci/localization-zh-cn-plugin/pull/31

Formerly downstream from https://github.com/kohsuke/localizer/pull/16

### CODE REVIEW

**Comment by daniel-beck** [comment]
This might be better off in the `localization` plugin, but the direction is unclear to me for further localizations.

**Comment by oleg-nenashev** [comment]
CC @jenkinsci/chinese-localization-sig @jonbrohauge

**Review by oleg-nenashev — COMMENTED** [review]
👍 for the approach. looking forward to get it merged

**Comment by LinuxSuRen** [comment]
LGTM, but we need to wait for the downstream.

**Comment by daniel-beck** [comment]
> LGTM, but we need to wait for the downstream.

Well, the proof of concept exists, so we _could_ already include this. That said, I want to get further with `localization-support` before introducing core changes that might be difficult to back out again.

**Comment by LinuxSuRen** [comment]
I'm just able to maintain the Simplified Chinese localization. AKA the *.zh_CN.* files.

**Comment by daniel-beck** [comment]
@LinuxSuRen That's fine. I removed `zh_TW` here only temporarily to make it easier to test and develop, that will be reverted.

**Comment by LinuxSuRen** [comment]
Got it.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Looks good to me overall. IIUC it will also need updates in the detached plugin list so that a proper version of https://github.com/jenkinsci/localization-zh-cn-plugin is referenced from there

**Comment by daniel-beck** [comment]
@oleg-nenashev It has never been considered split from core in the first place, I don't see a reason to do this now.

**Review by jvz — APPROVED** [review]

**Review by jeffret-b — APPROVED** [review]
Looks good once the conflict is resolved and we get another good build.

Looks like it's still waiting on an upstream dependency.

File: core/src/main/java/jenkins/MetaLocaleDrivenResourceProvider.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+package jenkins;
+
+import hudson.ExtensionList;
+import org.kohsuke.MetaInfServices;
+import org.kohsuke.stapler.LocaleDrivenResourceProvider;
+
+import java.net.URL;
+
+@MetaInfServices
+public final class MetaLocaleDrivenResourceProvider extends LocaleDrivenResourceProvider {
```
Comment (Wadeck) [inline core/src/main/java/jenkins/MetaLocaleDrivenResourceProvider.java]: `@Restricted(DoNotUse.class)` as it seems to be self-sufficient?

File: core/src/main/java/jenkins/MetaLocaleDrivenResourceProvider.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    @Override
+    public URL lookup(String s) {
+        /*
+         * This looks crazy but this instance only registers as MetaInfService so it can be looked up by Stapler.
+         * All other instances (in plugins) are expected to register as @Extension, making them available below.
+         *
+         * This saves having to think of one more stupid class name. The downside is the lack of "implements ExtensionPoint".
+         */
+        for (LocaleDrivenResourceProvider provider : ExtensionList.lookup(LocaleDrivenResourceProvider.class)) {
+            URL url = provider.lookup(s);
```
Comment (Wadeck) [inline core/src/main/java/jenkins/MetaLocaleDrivenResourceProvider.java]: nit: An additional layer of `try{...}catch` could be useful and is free in terms of perf.

**Review by Wadeck — APPROVED** [review]
Have you considered that all the new logic here will be called multiple times per page request and thus, could potentially impact the performance?

IIRC there is no code cache for the resource files / urls, and so, the addition/deletion of a localization plugins will have no impact on that, right?

WDYT about renaming `localization-support-plugin` to `localization-api-plugin` as it's not meant to be used directly?

**Comment by daniel-beck** [comment]
@Wadeck 

> Have you considered that all the new logic here will be called multiple times per page request and thus, could potentially impact the performance?

> IIRC there is no code cache for the resource files / urls, and so, the addition/deletion of a localization plugins will have no impact on that, right?

Not considered. Would expect this to not be worse than 2x the file system accesses, and that didn't seem too expensive. Note that Localizer does cache, and `localization-support` clears the cache as needed. Can always add further caching later, I think, if necessary. Notably, nothing here is really 

> WDYT about renaming `localization-support-plugin` to `localization-api-plugin` as it's not meant to be used directly?

No opinion. Both names indicate to me that they're not really immediately useful/standalone end-user features.

**Comment by daniel-beck** [comment]
Cleaned up the history, and focusing now on the localization-support infra (i.e. not removing localized files, too).

**Review by oleg-nenashev — APPROVED** [review]
Ready to go once Stapler is released and integrated

**Review by LinuxSuRen — APPROVED** [review]

**Comment by LinuxSuRen** [comment]
localizer 1.26 was released.

**Comment by daniel-beck** [comment]
@LinuxSuRen This needs a release of Stapler.

File: core/pom.xml [hunk 3]
```diff
@@ -39,7 +39,7 @@ THE SOFTWARE.
 
   <properties>
     <staplerFork>true</staplerFork>
-    <stapler.version>1.256</stapler.version>
+    <stapler.version>1.257-JEP-216-SNAPSHOT</stapler.version>
```
Comment (jglick) [inline core/pom.xml]: 1.257 (and next time you can use Incrementals!)

Comment (daniel-beck) [inline core/pom.xml]:
```suggestion
    <stapler.version>1.257</stapler.version>
```

**Review by jglick — CHANGES_REQUESTED** [review]

**Review by jglick — APPROVED** [review]
Better, thanks!

**Pull request merged**